### PR TITLE
Show all agents, smart sorted, in carousel on home screen of web app

### PR DESCRIPTION
**Pull request by debanjum**
## Overview
Allow quickly selecting, switching agents from agents pane on home page of web app

## Details
- Show all agents in carousel on home screen agent pane of web app
- Smart Sort
  1. Pin default agent as first for ease of access
  2. Show used agents by MRU for ease of access
  3. Shuffle unused agents for discoverability
- Select most recently used agent to chat with by default
- Push smart sort logic down to API
  - Common logic can be reused across clients
  - Agent sort was previously done in web app
- Focus on chat input on agent select
- Double click agent on home page to open edit agent card on agents page

## Preview

https://github.com/user-attachments/assets/a2701124-bb46-473f-896f-c8cdad6b7379

### CODE REVIEW

File: src/interface/web/app/components/chatInputArea/chatInputArea.tsx [hunk 1]
```diff
@@ -270,7 +254,7 @@ export default function ChatInputArea(props: ChatInputProps) {
         }
     }, [recording, mediaRecorder]);
 
-    const chatInputRef = useRef<HTMLTextAreaElement>(null);
+    const chatInputRef = ref as React.MutableRefObject<HTMLTextAreaElement>;
```
Comment (sabaimran) [inline src/interface/web/app/components/chatInputArea/chatInputArea.tsx:257]: Getting this error here after this change, consistently on older conversations as well. 

Screenshot 2024-10-22 at 12 00 39

**Review by sabaimran — CHANGES_REQUESTED** [review]

File: src/khoj/routers/api_agents.py [hunk 2]
```diff
@@ -45,30 +47,46 @@ async def all_agents(
...
+            agents_packet.append(agent_packet)
+
+    # Load Conversation Sessions
+    min_date = datetime.min.replace(tzinfo=timezone.utc)
+    conversations = []
+    if user:
+        conversations = await sync_to_async(list[Conversation])(
+            ConversationAdapters.get_conversation_sessions(user, request.user.client_app).reverse()
         )
+    conversation_times = {conv.agent.slug: conv.updated_at for conv in conversations}
```
[the remark this answers is not in the document]
Comment (sabaimran) [reply src/khoj/routers/api_agents.py:82]:
```suggestion
    conversation_times = {}
    for conv in conversations:
        if conv.agent and conv.agent.slug:
            conversation_times[conv.agent.slug] = conv.updated_at
```

File: src/interface/web/app/page.tsx [hunk 3]
```diff
@@ -194,34 +207,73 @@ function ChatBodyData(props: ChatBodyDataProps) {
...
-                                        : "border-stone-100 dark:border-neutral-700 text-muted-foreground"
-                                }
-                                    hover:cursor-pointer rounded-lg px-2 py-2`}
-                            >
-                                <CardTitle
-                                    className="text-center text-md font-medium flex justify-center items-center"
-                                    onClick={() => setSelectedAgent(agents[index].slug)}
+                    <ScrollArea className="w-full max-w-[600px] mx-auto">
+                        <div className="flex pb-2 gap-2 items-center justify-center">
+                            {agents.map((agent, index) => (
```
[the remark this answers is not in the document]
Comment (debanjum) [reply src/interface/web/app/page.tsx:212]: - [X] Removed the border from agent details hover card
- [X] Made long agent names not wrap in agent pill
- [ ] There is logic put in to make the hover card disappear the moment the cursor is not on the card. But I think scroll away doesn't trigger that. But once the scroll stops it triggers. It is a minor annoyance. But I'll let this be for now

File: src/interface/web/app/components/chatInputArea/chatInputArea.tsx (hunk 1 above)
Comment (debanjum) [reply src/interface/web/app/components/chatInputArea/chatInputArea.tsx:257]: Ouch. Fixed that by adding a null check for chatInputRef as well. Let me know if you're still seeing this issue?

File: src/khoj/routers/api_agents.py [hunk 4]
```diff
@@ -45,30 +47,46 @@ async def all_agents(
...
+            default_agent_packet = agent_packet
+        else:
+            agents_packet.append(agent_packet)
+
+    # Load Conversation Sessions
+    min_date = datetime.min.replace(tzinfo=timezone.utc)
+    conversations = []
+    if user:
+        conversations = await sync_to_async(list[Conversation])(
+            ConversationAdapters.get_conversation_sessions(user, request.user.client_app).reverse()
```
[the remark this answers is not in the document]
Comment (debanjum) [reply src/khoj/routers/api_agents.py:80]: Good point. Improved that to only load upto 50 conversation sessions from the last 2 weeks to compute most recently used agents

File: src/khoj/routers/api_agents.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (debanjum) [reply src/khoj/routers/api_agents.py:82]: Nice catch, didn't think about older conversations with no associated agents with it. Fixed it to check for conv.agent being present (Assumes if conv.agent is set then conv.agent.slug shouldn't be null). Let me know if you're still seeing this issue after the fix

File: src/khoj/routers/api_agents.py [hunk 5]
```diff
@@ -45,30 +47,46 @@ async def all_agents(
...
+
+    # Put default agent first, then sort by mru and finally shuffle unused randomly
+    random.shuffle(agents_packet)
+    agents_packet.sort(key=lambda x: conversation_times.get(x["slug"]) or min_date, reverse=True)
+    if default_agent_packet:
+        agents_packet.insert(0, default_agent_packet)
 
-    # Make sure that the agent named 'khoj' is first in the list. Everything else is sorted by name.
-    agents_packet.sort(key=lambda x: x["name"])
-    agents_packet.sort(key=lambda x: x["slug"] == "khoj", reverse=True)
```
[the remark this answers is not in the document]
Comment (debanjum) [reply src/khoj/routers/api_agents.py:71]: Yeah, this logic does keep the default agent first in the list

**Review by sabaimran — APPROVED** [review]
🎉

**Pull request merged**